packages/accounts/src/light-account/account.ts

```diff
@@ -55,6 +65,92 @@ export class LightSmartContractAccount<
     return decodedCallResult;
   }
 
+  /**
+   * Upgrades the account implementation from Light Account to a Modular Account.
+   * Optionally waits for the transaction to be mined.
+   *
+   * @param provider - the provider to use to send the transaction
+   * @param chain - the chain to upgrade the account on
+   * @param accountImplAddress - the address of the smart account implementation to
+   * upgrade to
+   * @param initializationData - the initialization data address to use when upgrading to the new
+   * smart account
+   * @param waitForTxn - whether or not to wait for the transaction to be mined
+   * @returns {
+   *  provider: SmartAccountProvider<TTransport> & { account: MSCA };
+   *  hash: Hash;
+   * } - the upgraded provider and corresponding userOperation hash,
+   * or transaction hash if `waitForTxn` is true
+   */
+  static async upgrade<
```

Comment:
This is why making the upgrade function generic is difficult. the return type depends on the account type introduced to create a "connected" provider via `.connect`. 

We ran into the same issue with the `createLightAccountAlchemyProvider` factory PR. 

Any ideas on how to genericize without running into this error, which can only be resolved with an unknown cast:
![image](https://github.com/alchemyplatform/aa-sdk/assets/43521356/dd68ef45-f406-4a68-8d5e-bf62fd80cb03)
![Screenshot 2023-11-30 at 8 09 55 PM](https://github.com/alchemyplatform/aa-sdk/assets/43521356/63559285-bc30-4fad-b150-382e8767f905)

Reply:
oh I didn't even think about the fact that reconnecting a provider generates a conflict of types... which makes sense because connect is `connect<A>(acct: A) => this & { account: A}` so if you do `connect` when already connected you get `this & { account: A} & {account: B }`